numpy/ma/core.py

```diff
@@ -4287,22 +4298,44 @@ def count(self, axis=None):
         array([3, 0])
 
         """
+        kwargs = {}
+        if keepdims is not np._NoValue:
+            kwargs['keepdims'] = keepdims
+
         m = self._mask
-        s = self.shape
-        if m is nomask:
+        # optimization if we exactly have an ndarray
+        if m is nomask and type(self._data) is np.ndarray:
+            # see _count_reduce_items in _methods.py
+            axes = axis
             if axis is None:
-                return self.size
+                axes = tuple(range(self.ndim))
+            if not isinstance(axes, tuple):
+                axes = (axis,)
+
+            items = 1
+            for ax in axes:
+                items *= self.shape[ax]
+
+            if self.shape is () or axis is None:
+                return items
+            elif kwargs.get('keepdims', False):
+                out_dims = list(self.shape)
+                for a in axes:
+                    out_dims[a] = 1
             else:
-                n = s[axis]
-                t = list(s)
-                del t[axis]
-                return np.full(t, n, dtype=np.intp)
-        n1 = np.size(m, axis)
-        n2 = np.sum(m, axis=axis, dtype=np.intp)
-        if axis is None:
-            return (n1 - n2)
-        else:
-            return narray(n1 - n2)
+                out_dims = [d for n,d in enumerate(self.shape) if n not in axes]
+            return np.asarray(np.intp(items)*ones(out_dims, dtype=np.intp))
```

Comment:
That doesn't quite work either because it can return a scalar if `out_dims == ()` (we need to always return a 0-d array here, according to unit tests), but you're right the code can be improved. I'll fix it.